packages/service-provider/src/lib/service-provider.ts

```diff
@@ -0,0 +1,304 @@
+      }
+
+      const protocolSlug = parsedSST.payload.network;
+
+      const { permit, permitSig } = parsedSST.payload;
+      const nowSeconds = Math.floor(Date.now() / 1000);
+      if (Number(permit.deadline) > nowSeconds) {
+        // Retrieve owner from ERC721 contract
+        const { tokenId } = permit.permitted;
+        const tokenIdAsStr = tokenId.toString();
+
+        const { smartAssetContract } = await this.getProtocolClientInstance(
+          protocolSlug
+        );
+
+        const owner = await smartAssetContract.ownerOf(tokenIdAsStr);
+        // Check if owner is the same as the one who signed the permit
+        if (owner !== parsedSST.payload.iss) {
+          throw new Error(
+            `Owner address ${owner} is not the same as the one who signed the permit ${parsedSST.payload.iss}`
+          );
+        }
+
+        // Check if owner has approved Permit721 contract to transfer the token
+        const approvedAddress = await smartAssetContract.getApproved(
+          tokenIdAsStr
+        );
+        const isApproved =
+          approvedAddress && approvedAddress === PERMIT721_ADDRESS;
+        if (!isApproved)
+          throw new Error(
+            `Owner ${owner} has not approved Permit721 contract ${PERMIT721_ADDRESS}`
+          );
+
+        if (performDryRun) {
+          // Simulate the transfer through Permit721 contract
+          await this._transferSmartAsset(
+            owner,
+            tokenIdAsStr,
+            permit,
+            permitSig,
+            DRY_RUN_ADDRESS,
+            protocolSlug,
+            true
+          );
+        }
+
+        return {
+          valid: true,
+          owner,
+          tokenId: tokenIdAsStr,
+          permit,
+          permitSig,
+          protocolSlug,
+        };
+      } else {
+        throw new Error(
+          `SST is expired (deadline: ${permit.deadline}, now: ${nowSeconds})`
+        );
+      }
+    } catch (err) {
+      console.error(`An error occurred while validating SST: ${err} (${sst})`);
+      return {
+        valid: false,
+        owner: undefined,
+        tokenId: undefined,
+        permit: undefined,
+        permitSig: undefined,
+      };
+    }
+  }
+
+  public async getSmartAssetFromSST({
+    sst,
+    performDryRun = false,
+  }: {
+    sst: string;
+    performDryRun?: boolean;
+  }) {
+    const { valid, tokenId, protocolSlug } = await this._isValidSST(
+      sst,
+      performDryRun
+    );
+    if (!valid) {
+      throw new Error(`SST ${sst} is not valid`);
+    }
+
+    const core = Core.fromRandom(); // We use a random core here, as we don't need to sign anything
+    const aat = new ArianeeAccessToken(core, {
+      initialValues: {
+        walletAccessToken: sst,
+      },
+    });
+
+    const walletApiClient = new WalletApiClient(
+      getChainTypeOf(protocolSlug!),
+      core,
+      {
+        arianeeAccessToken: aat,
+      }
+    );
+    return walletApiClient.getSmartAsset(protocolSlug!, {
```

Comment:
Why not using the Wallet lib instead, so we get the smartAsset content

Reply:
yes might be a better option, so that we get the arianee events too (in a `SmartAssetInstance`) (which might be a requirement for displaying the product on the service provider website)